### PR TITLE
[PART 1] Harden Socket

### PR DESCRIPTION
**Pull request by connor4312**
A bit of a notable diff here. I noticed several bugs with Carina behaviour on unstable connections and with reconnections when developing locally. This is the first part of a series of PRs which focuses on hardening the ConstellationSocket. Subsequent PRs will look at getting pubsub events reliable and the higher-level Carina interface up to date.

In general I redid the data system of the Socket class. In bullet-point form:
- We now only provide methods to execute RPC calls over Constellation, getting rid of `sendJson`. The only type of data sent from clients to the server are RPC calls.
- The config object used to instantiate the socket has been adjusted. Notably:
  - the socket now requires a specific connect() call, rather than connecting immediately upon instantiation
  - the gzip detection is provided as an interface rather than a plain numeric byte threshold, which is needed for some secret stuff on frontend
  - the reconnection policy is configurable and is also provided as an interface
- states have been adjusted to be a bit less error-prone than multiple booleans which were able to get into a bad state
- all packets are now stored in the queue until they are cancelled or completed; this protects against disconnections which may occur in the middle of method calls
- timeouts are now measured from the time the packet is actually sent rather than from the time `execute()` is called
- fixed some other misc bugs; bad states occurring after more than one disconnection occurred, user agent being omitted, the ability to set both jwt and oauth tokens for auth, etc.
- unit tests!

### CODE REVIEW

File: src/packets.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    // The packet has been sent over the websocket successfully and we are
+    // waiting for a reply.
+    Sending,
+    // The packet was replied to, and has now been complete.
+    Replied,
+    // The caller has indicated they no longer wish to be notified about this event.
+    Cancelled
+}
+
+const maxInt32 = -1 >>> 0;
```
Comment (SimonSchick) [inline src/packets.ts]: I feel like `0xFFFFFFFF` would feel better rather than some magic bitshift.

File: src/util.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                if (!done) {
+                    done = true;
+                    fn(value);
+                }
+            };
+        }
+
+        const resolve = guard(_resolve);
+        const reject = guard(_reject);
+        promises.forEach(p => p.then(resolve).catch(reject));
```
Comment (SimonSchick) [inline src/util.ts]: Shouldn't this be `.then(resolve, reject)`?

File: test/socket.test.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            socket = new Socket({ url, authToken: 'asdf!' }).connect();
+            server.on('connection', ws => {
+                expect(ws.upgradeReq.url).to.equal('/');
+                expect(ws.upgradeReq.headers.authorization).to.equal('Bearer asdf!');
+                done();
+            });
+        });
+
+        it('throws an error on ambiguous auth', () => {
+            expect(() => new Socket({ url, authToken: 'asdf!', jwt: 'wat?' })).to.throw();
```
Comment (SimonSchick) [inline test/socket.test.js]: to throw what?

Comment (SimonSchick) [inline test/socket.test.js]: To clarify, the fact that something throws doesn't mean it throws the right thing.

**Comment by SimonSchick** [comment]
Looks good overall, I might have commented on code segments you moved tho.

File: test/socket.test.js (hunk 3 above)
Comment (connor4312) [inline test/socket.test.js]: Yep that's fair

File: src/socket.ts [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    // the socket is connection and data may be sent
+    Connected,
+    // the socket is gracefully closing; after this it will become Idle
+    Closing,
+}
+
+function getDefaults(): SocketOptions {
+    return {
+        url: 'wss://constellation.beam.pro',
+        userAgent: `Carnia ${pkg.version}`,
```
Comment (jamesbirtles) [inline src/socket.ts]: Carina

File: src/util.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import { TimeoutError } from './errors';
+
+/**
+ * race takes an array of promises and is resolved with the first promise
+ * that resolves, or rejects. After the first resolve or rejection, all
+ * future resolutions and rejections will be discarded.
+ * @param {Promise[]} promises
+ * @return {Promise}
+ */
+export function race (promises: Promise<any>[]): Promise<any> {
```
Comment (jamesbirtles) [inline src/util.ts]: Why not just use Promise.race?

Comment (connor4312) [inline src/util.ts]: Wow I totally missed that that was a thing. Will change :P

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export * from './carina';
```
Comment (SimonSchick) [inline src/index.ts:1]: I never really good the point of files that that aggregate other files and then export them, why not leave it up the the end user what they want to import?

File: src/packets.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * Returns the packet's timeout duration, or the default if undefined.
+     */
+    getTimeout(defaultTimeout: number): number {
+        return this.timeout || defaultTimeout;
+    }
+
+    /**
+     * Returns the current state of the packet.
+     * @return {PacketState} [description]
```
Comment (SimonSchick) [inline src/packets.ts]: rm doc placeholder

File: src/packets.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            id: Math.floor(Math.random() * maxInt32),
+            type: 'method',
+            method,
+            params,
+        };
+    }
+
+    /**
+     * Returns the randomly-assigned numeric ID of the packet.
+     * @return {number}
```
Comment (SimonSchick) [inline src/packets.ts:42]: number or Number?
You should with one, right now it's mixed.

File: src/packets.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * Returns the current state of the packet.
+     * @return {PacketState} [description]
+     */
+    getState(): PacketState {
+        return this.state;
+    }
+
+    setState(state: PacketState) {
+        if (state == this.state) {
```
Comment (SimonSchick) [inline src/packets.ts]: Why no strict equal?

File: src/index.ts (hunk 6 above)
Comment (jamesbirtles) [reply src/index.ts:1]: Because the consumer should not care about the internal structure of the module.

**Comment by SimonSchick** [comment]
👍

File: src/packets.ts (hunk 8 above)
Comment (connor4312) [reply src/packets.ts:42]: For JsDoc, there is no difference. For TypeScript, it always should be lowercase `number` as number types aren't instances of Number. (Yay JavaScript). Soo, whichever :P

**Pull request merged**